File: 0155-min-stack/0155-min-stack.py
class Stack():
    def __init__(self):
        self._a = []
        self._max_size = 0

    def push(self, x: int) -> None:
        self._a.append(x)
        self._max_size = max(self._max_size, len(self._a))

    def pop(self) -> int:
        return self._a.pop() if self._a else None

    def top(self) -> int:
        return self._a[-1] if self._a else None

    def empty(self) -> bool:
        return len(self._a) == 0

    def __len__(self) -> int:
        return len(self._a)

    def space(self) -> int:
        return self._max_size
        
class MinStack():
    def __init__(self):
        self.stack = Stack()
        self.minStack = Stack()

    def push(self, x: int) -> None:
        self.stack.push(x)
        if self.minStack.empty() or x <= self.minStack.top():
            self.minStack.push(x)

    def pop(self) -> None:
        if self.stack.top() == self.minStack.top():
            self.minStack.pop()
        self.stack.pop()

    def top(self) -> int:
        return self.stack.top()

    def getMin(self) -> int:
        return self.minStack.top()        
        

if (True):
    MinStack = MinStack

File: 0155-min-stack/0155-min-stack.java

// Here minStack is monotonically decreasing from bottom to top.

class MinStack {
    private Stack<Integer> stack;
    private Stack<Integer> minStack;

    public MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }
    
    public void push(int val) {
        // Push the value onto the main stack
        stack.push(val);
        
        // Push onto the minStack if it's empty or the new value is the current minimum
        if (minStack.isEmpty() || val <= minStack.peek()) {
            minStack.push(val);
        }
    }
    
    public void pop() {
        // Pop from the main stack
        int poppedValue = stack.pop();
        
        // If the popped value is the minimum, pop it from minStack too
        if (poppedValue == minStack.peek()) {
            minStack.pop();
        }
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minStack.peek();
    }
}


/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */